# **Hands-on 1: Clasificador Lineal, Perceptron**

**Clasificación Inteligente de Datos**

Daniel García Figueroa

## **1.1  Fundamentos de la técnica**

El perceptrón es una neurona artificial creada en 1957, que representa la forma más simple de red neuronal. Funciona como un algoritmo de aprendizaje supervisado diseñado principalmente para clasificación binaria, es decir, para separar datos en dos categorías distintas, como determinar si algo es 0 o 1.

---

### **¿Cómo funciona?**

 Su funcionamiento es el siguiente:
 1. Recibe varias entradas de datos
 2. Multiplica cada una por un peso que indica su importancia
 3. Suma todos estos valores y compara el resultado con un umbral establecido.

 Si la suma supera ese umbral, la neurona se activa y da como salida 1, de lo contrario produce 0. Durante el entrenamiento, cuando se equivoca en sus predicciones, el sistema ajusta automáticamente los pesos para mejorar su precisión. Este proceso de aprendizaje le permite detectar patrones en los datos y hacer clasificaciones cada vez más exactas.


---

### **Tipos**

 Hay dos tipos de perceptrón: con **capa única** y **multicapas**.
 - Un perceptrón con capa única puede enseñar únicamente funciones lineales separables.
 - Un perceptrón multicapas, también llamado red neuronal “feed-forward” permite superar ese límite y ofrece una potencia de cálculo superior. También es posible combinar varios perceptrones para crear un mecanismo poderoso.

## **1.2 Modelo Matemático del Perceptrón**

El perceptrón es uno de los modelos más básicos de redes neuronales y se basa en una combinación lineal de las entradas, seguida por una función de activación que decide la clase de salida.  

Matemáticamente, el conjunto de datos se puede representar como:

$$
D = \{ (\mathbf{x}_1, y_1), (\mathbf{x}_2, y_2), \dots, (\mathbf{x}_n, y_n) \}
$$

$$
\mathbf{x}_i \in \mathbb{R}^d \text{ representa un vector de características, y } y_i \in \{0, 1\} \text{ es la etiqueta asociada a cada muestra.}
$$

---

### **Combinación Lineal**

El perceptrón calcula una suma ponderada de las entradas junto con un sesgo:

$$
z = \mathbf{w} \cdot \mathbf{x} + b
$$

donde:  
- (w) son los **pesos** del modelo,  
- (x) es el **vector de entrada**,  
- (b) es el **sesgo (bias)**.

---

### **Función de Activación**

La salida del perceptrón se obtiene aplicando una **función escalón** al resultado anterior:

$$
\hat{y} =
\begin{cases}
1, & \text{si } (\mathbf{w} \cdot \mathbf{x} + b) \geq 0 \\
0, & \text{si } (\mathbf{w} \cdot \mathbf{x} + b) < 0
\end{cases}
$$

---

### **Regla de Actualización de Pesos**

Durante el entrenamiento, los pesos se ajustan cada vez que el modelo comete un error. La actualización se realiza según:

$$
\mathbf{w} = \mathbf{w} + \eta (y_{\text{real}} - y_{\text{pred}}) \mathbf{x}
$$

$$
b = b + \eta (y_{\text{real}} - y_{\text{pred}})
$$

$$\eta \text{ es la tasa de aprendizaje, y } (y_{\text{real}} - y_{\text{pred}}) \text{ representa el error de predicción.}
$$

Este proceso iterativo continúa hasta que el modelo logra clasificar correctamente los datos linealmente separables.


## **1.3  Descripción de la librería, clases, funciones en su respectiva sección.**



## **1.4 Pipeline**

El objetivo de implementar el modelo Perceptrón es tratar de predecir si una persona tiene o no diabetes a partir de distintas características médicas incluidas en el dataset.

El propósito principal del modelo es aprender a distinguir entre los pacientes diabéticos y los no diabéticos. El perceptrón ajusta sus pesos de manera iterativa durante el entrenamiento para minimizar los errores de clasificación.

### **1.4.1 Feature Engineering**


En el dataset de diabetes que estoy utilizando hay 8 variables que describen distintas características de cada paciente. La variable objetivo "Outcome" indica si la persona tiene o no diabetes.

1. Embarazos
2. Glucosa
3. Presión arterial
4. Grosor de piel
5. Insulina
6. Indice de masa corporal
7. Historial familiar de diabetes
8. Edad


In [2]:
import pandas as pd # Está diseñada específicamente para la manipulación y el análisis de datos en el lenguaje Python. Trabaja sobre dataframes

df = pd.read_csv('/content/sample_data/diabetes.csv') # Se lee el archivo de csv del dataset que subí manualmente

# Muestra estadísticas descriptivas del dataset
print("Resumen estadístico:")
print(df.describe().round(2))
print()
# Muestra el tamaño del dataset (filas, columnas)
print("Tamaño del dataset (filas,columnas):")
print(df.shape)
print()
# Muestra las primeras 5 filas
print("Primeras 5 filas:")
print(df.head())
print()


Resumen estadístico:
       Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin     BMI  \
count       768.00   768.00         768.00         768.00   768.00  768.00   
mean          3.85   120.89          69.11          20.54    79.80   31.99   
std           3.37    31.97          19.36          15.95   115.24    7.88   
min           0.00     0.00           0.00           0.00     0.00    0.00   
25%           1.00    99.00          62.00           0.00     0.00   27.30   
50%           3.00   117.00          72.00          23.00    30.50   32.00   
75%           6.00   140.25          80.00          32.00   127.25   36.60   
max          17.00   199.00         122.00          99.00   846.00   67.10   

       DiabetesPedigreeFunction     Age  Outcome  
count                    768.00  768.00   768.00  
mean                       0.47   33.24     0.35  
std                        0.33   11.76     0.48  
min                        0.08   21.00     0.00  
25%                  

### **1.4.2 Modelo Selection**

Decidí utilizar un clasificador lineal porque el Perceptrón es un modelo sencillo pero eficaz para problemas de clasificación binaria, como el del dataset de diabetes (donde el resultado solo puede ser si tiene o no diabetes).

Este tipo de modelo permite ver cómo cada variable influye en la decisión, ya que asigna un peso específico a cada característica. Esto ayuda a entender qué factores son más importantes para predecir la diabetes.

### **1.4.3 Model Training**


In [14]:
import numpy as np # Librería para operaciones numéricas y conteos
from sklearn.model_selection import train_test_split # Librería para dividir datos en entrenamiento y prueba
from sklearn.linear_model import Perceptron # Librería del modelo a utilizar
from sklearn.preprocessing import StandardScaler # Librería para normalizar los datos

# Elimino la variable objetivo porque es lo que queremos predecir
X = df.drop('Outcome', axis=1).values
# Se guardan los valores de la columna objetivo
y = df['Outcome'].values

# Separa los datos en conjunto de entrenamiento y prueba (80% - 20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Normalización de las características
scaler = StandardScaler() # Crea un objeto StandardScaler para escalar los datos
X_train_scaled = scaler.fit_transform(X_train)   # Ajuste y transformación con datos de entrenamiento
X_test_scaled = scaler.transform(X_test)         # Solo transformación con datos de prueba


print(f"Datos de entrenamiento: {X_train.shape}") # Cantidad de filas y columnas en entrenamiento
print(f"Datos de prueba: {X_test.shape}") # Cantidad de filas y columnas en prueba
print()
print(f"Distribución en entrenamiento:")
print(f"  No Diabetes (0): {np.sum(y_train==0)}") # Cuenta de casos sin diabetes
print(f"  Diabetes (1): {np.sum(y_train==1)}") # Cuenta de casos con diabetes
print()
print(f"Distribución en prueba:")
print(f"  No Diabetes (0): {np.sum(y_test==0)}") # Cuenta de casos sin diabetes
print(f"  Diabetes (1): {np.sum(y_test==1)}") # Cuenta de casos con diabetes
print()

# Entrenamiento del perceptrón solo con los datos de entrenamiento
perceptron = Perceptron(max_iter=1000, eta0=0.01, random_state=42, class_weight='balanced') # El parámetro class_weight='balanced' ajusta los pesos automáticamente para compensar el desbalance entre clases
perceptron.fit(X_train_scaled, y_train) # Se entrena el perceptrón usando los datos de entrenamiento escalados

# Imprimir los pesos y el bias aprendidos
print("Pesos aprendidos:", perceptron.coef_)
print("Bias aprendido:", perceptron.intercept_)

Datos de entrenamiento: (614, 8)
Datos de prueba: (154, 8)

Distribución en entrenamiento:
  No Diabetes (0): 401
  Diabetes (1): 213

Distribución en prueba:
  No Diabetes (0): 99
  Diabetes (1): 55

Pesos aprendidos: [[-0.00186671  0.06651944 -0.04281165  0.0180872   0.00159093  0.05679316
   0.02991816 -0.00866989]]
Bias aprendido: [-0.03155796]


### **1.4.4 Prediction**


In [23]:
# Función para predecir la clase de una entrada
def probar_patron(perceptron, X): # Recibe el perceptrón entrenado anteriormente con los datos de entrenamiento y un conjunto de datos
    y_pred = perceptron.predict(X) # Almacena las predicciones de si tiene o no diabetes
    return y_pred # Retorna la predicción

In [25]:
# Recorre 30 registros de mi conjunto de pruebas
for i in range(30):
    y_pred = probar_patron(perceptron, [X_test_scaled[i]]) # Usamos la función declarada arriba, pasándole el perceptrón y la instancia del conjunto de prueba del recorrido
    print(f"Instancia {i+1}: {X_test_scaled[i].round(2)}, Salida Esperada: {y_test[i]}, Salida Predicha: {y_pred[0]}")

Instancia 1: [ 0.68 -0.71 -0.62  0.82  0.93  0.26 -0.12  0.88], Salida Esperada: 0, Salida Predicha: 0
Instancia 2: [-0.53 -0.28  0.3   0.75 -0.7   0.48 -0.95 -1.04], Salida Esperada: 0, Salida Predicha: 0
Instancia 3: [-0.53 -0.4  -0.29 -1.32 -0.7  -0.15 -0.92 -1.04], Salida Esperada: 0, Salida Predicha: 0
Instancia 4: [ 1.29 -0.43  0.57 -1.32 -0.7  -0.95  1.15  0.1 ], Salida Esperada: 0, Salida Predicha: 0
Instancia 5: [ 0.98  0.47  1.11 -1.32 -0.7  -0.27 -0.77  1.49], Salida Esperada: 0, Salida Predicha: 0
Instancia 6: [ 0.68 -0.56  0.14  0.75  0.93  0.74 -0.43  1.92], Salida Esperada: 0, Salida Predicha: 0
Instancia 7: [-0.83 -1.56 -1.16 -0.16 -0.05 -1.5  -0.43 -0.95], Salida Esperada: 0, Salida Predicha: 0
Instancia 8: [-1.13 -0.12 -3.75 -1.32 -0.7   0.23  1.38  0.97], Salida Esperada: 0, Salida Predicha: 1
Instancia 9: [ 0.08  1.04  0.14  0.56  0.38 -0.09 -0.39  0.36], Salida Esperada: 0, Salida Predicha: 1
Instancia 10: [ 0.38  0.82  0.46 -1.32 -0.7   0.22 -0.75  2.79], Salida E

### **1.4.5 Model Evaluation**


In [26]:
from sklearn.metrics import accuracy_score # Función para calcular la precisión del modelo

y_pred = perceptron.predict(X_test_scaled) # Usa el perceptrón entrenado para predecir las salidas de X de prueba
print("Accuracy:", accuracy_score(y_test, y_pred) ) # Muesta la precisión del modelo

Accuracy: 0.7402597402597403


## **Conclusión**
El modelo que implementé logró una precisión del 74.03% en el conjunto de prueba, lo que significa que más o menos 7 de cada 10 predicciones del perceptrón fueron correctas. Al principio, el accuracy estaba cerca del 61%, pero al normalizar los datos mejoró bastante. También jugué un poco con los parámetros del perceptrón, como el número de iteraciones y la tasa de aprendizaje, aunque eso no cambió demasiado el resultado. Creo que es un desempeño razonable y, hasta ahora, la mejor precisión que pude obtener. Pienso que el desbalanceo entre clases influye bastante, porque hay casi el doble de casos de personas no diabéticas que de personas diabéticas.

## **Referencias bibliográficas**

Daniel. (2023, October 30). Perceptrón: ¿qué es y para qué sirve? DataScientest. https://datascientest.com/es/perceptron-que-es-y-para-que-sirve

GeeksforGeeks. (2025, July 23). What is Perceptron | The Simplest Artificial neural network. GeeksforGeeks. https://www-geeksforgeeks-org.translate.goog/machine-learning/what-is-perceptron-the-simplest-artificial-neural-network/?_x_tr_sl=en&_x_tr_tl=es&_x_tr_hl=es&_x_tr_pto=tc

López, R. (2025, September 22). Mathematics of the perceptron neuron model. Neural Designer. https://www.neuraldesigner.com/blog/perceptron-the-main-component-of-neural-networks/

Garau, G., & Garau, G. (2025, October 21). El Perceptrón Simple: Implementación en Python. Damavis Blog - Data - Machine Learning - Visualization. https://blog.damavis.com/el-perceptron-simple-implementacion-en-python/#ejemplo-de-implementacion-en-python

DataScientest. (2023, October 30). Pandas : La biblioteca de Python dedicada a la Data Science. https://datascientest.com/es/pandas-python